docs/RFCS/20210526_expression_indexes.md

```diff
@@ -0,0 +1,288 @@
+
+To address these challenges, when displaying any index, an indexed virtual
+column's expression is displayed rather than its name if:
+
+1. It has a name with the prefix `crdb_idx_expr`.
+2. It is a hidden virtual column.
+
+When the table containing an expression index is displayed, the hidden virtual
+column(s) that the index relies on is displayed as `NOT VISIBLE`. This is
+standard for all hidden columns.
+
+The `crdb_idx_expr` column name prefix communicates that a column originated
+from an expression index. This is critical for satisfying **C**. It must be
+communicated that a column was created for an expression index to avoid creating
+a duplicate column when the output of `SHOW CREATE TABLE` is executed. The
+output of `SHOW CREATE TABLE` cannot communicate arbitrary information about
+columns; it must be valid SQL.
+
+## Putting it all together
+
+The following example shows the outcome of creating, dropping, and displaying
+expression indexes.
+
+```sql
+CREATE TABLE t (
+    k INT PRIMARY KEY,
+    a INT,
+    b INT,
+    INDEX t_a_plus_b_idx ((a + b))
+);
+
+SELECT create_statement FROM [SHOW CREATE TABLE t];
+                  create_statement
+------------------------------------------------------
+CREATE TABLE public.t (
+    k INT8 NOT NULL,
+    a INT8 NULL,
+    b INT8 NULL,
+    crdb_idx_expr INT8 NOT VISIBLE NULL AS (a + b) VIRTUAL,
+    CONSTRAINT "primary" PRIMARY KEY (k ASC),
+    INDEX t_a_plus_b_idx ((a + b) ASC)
+);
+
+DROP INDEX t_a_plus_b_idx;
+
+SELECT create_statement FROM [SHOW CREATE TABLE t];
+                  create_statement
+------------------------------------------------------
+CREATE TABLE public.t (
+    k INT8 NOT NULL,
+    a INT8 NULL,
+    b INT8 NULL,
+    CONSTRAINT "primary" PRIMARY KEY (k ASC)
+);
+```
+
+The following example shows the behavior when executing the output of `SHOW
+CREATE TABLE` of a table with an expression index. Notice that an identical
+table is created. The `crdb_idx_expr` hidden virtual column is not duplicated
+because the expression index uses it rather than creating a new column.
+
+```sql
+CREATE TABLE public.t (
+    k INT8 NOT NULL,
+    a INT8 NULL,
+    b INT8 NULL,
+    crdb_idx_expr INT8 NOT VISIBLE NULL AS (a + b) VIRTUAL,
+    CONSTRAINT "primary" PRIMARY KEY (k ASC),
+    INDEX t_a_plus_b_idx ((a + b) ASC)
+);
+
+SELECT create_statement FROM [SHOW CREATE TABLE t];
+                  create_statement
+------------------------------------------------------
+CREATE TABLE public.t (
+    k INT8 NOT NULL,
+    a INT8 NULL,
+    b INT8 NULL,
+    crdb_idx_expr INT8 NOT VISIBLE NULL AS (a + b) VIRTUAL,
+    CONSTRAINT "primary" PRIMARY KEY (k ASC),
+    INDEX t_a_plus_b_idx ((a + b) ASC)
+);
+```
+
+## Drawbacks
+
+### Relying on the name of virtual columns is brittle
+
+A user could create a hidden virtual column with a name prefixed with
+`crdb_idx_expr`. They could then create an index on this column. If the index
```

Comment:
My feeling is that we should (1) make the prefix `crdb_internal_idx_expr_...` and (2) ban users from creating columns (or honestly anything) that has the prefix `crdb_internal`. 

The day somebody comes and makes a github issue saying that that breaks their existing application is the day I tell that person to fix that application and close the issue.

Reply:
(1) sounds good to me.

(2) makes sense, but not sure it should be part of this RFC. It's probably worth careful consideration of the impact of this. And worth the time to think of the _anything's_ that should be banned from using this name.